source/math/FloatVectorList.ooc

```diff
@@ -381,4 +381,85 @@ FloatVectorList: class extends VectorList<Float> {
 			result add(this[i] clamp(floor, ceiling))
 		result
 	}
+	findMaxIndex: func -> Int {
+		index := 0
+		for (i in 1 .. this count)
+			if (this[index] < this[i])
+				index = i
+		index
+	}
+	crossCorrelation: func (other: This, vectorLength: Int) -> Float {
+		upperPart := 0.0f
+		lowerPart1 := 0.0f
+		lowerPart2 := 0.0f
+		thisMean := this mean
+		otherMean := other mean
+		thisPointer := (this pointer as Float*)
+		otherPointer := (other pointer as Float*)
+		for (i in 0 .. vectorLength) {
+			upperPart += ((thisPointer[i] - thisMean) * (otherPointer[i] - otherMean))
+			lowerPart1 += ((thisPointer[i] - thisMean) * (thisPointer[i] - thisMean))
+			lowerPart2 += ((otherPointer[i] - otherMean) * (otherPointer[i] - otherMean))
+		}
+		upperPart / sqrt(lowerPart1 * lowerPart2)
+	}
+	shift: func (offset: Int, filledValue := 0.0f) -> This {
+		result := This new(this count)
+		thisPointer := (this pointer as Float*)
+		if (offset < 0)
+			for (i in 0 .. this count)
+				if ((i - offset) <= this count)
+					result add(thisPointer[i - offset])
+				else
+					result add(filledValue)
+		else
+			for (i in 0 .. this count)
+				if (i < offset)
+					result add(filledValue)
+				else
+					result add(thisPointer[i - offset])
+		result
+	}
+	// offset < 0 means it shifts to left
+	findCrossCorrelationOffset: func (other: This, offsetStart, offsetEnd: Int) -> CorrelationData {
+		crossCorrelation := This new(this count)
+		for (offset in offsetStart .. offsetEnd)
+			crossCorrelation add(this shift(offset) crossCorrelation(other, other count))
+
+		index := crossCorrelation findMaxIndex()
+		maxValue := this[index]
+		index = (index < (offsetEnd - offsetStart) / 2) ? (index + offsetStart) : (index - offsetEnd)
+
+		CorrelationData new(index, maxValue, crossCorrelation)
+	}
+	//input: how many points there will be between two origin points.
+	// this is a Linear interpolation way to have higher precision
+	higherPrecision: func (numberOfPoints: Int) -> This {
+		if (numberOfPoints != 0) {
+			newCount := this count + numberOfPoints * (this count - 1)
+			result := This new(newCount)
+			precision := (numberOfPoints + 1) as Float
+			thisPointer := (this pointer as Float*)
+			for (index in 0 .. newCount) {
```

Comment:
It will be in a separate PR for sure. I'm just wondering if we need to do that or not.